### PR TITLE
Fix BatchBALD, now same results as official impl

### PR DESCRIPTION
**Pull request by Dref360**
## Summary:

The issue was that we were not sampling from the history correctly.	

I have tested this version against the official one: https://github.com/BlackHC/BatchBALD and it returns the same indices.

### CODE REVIEW

File: src/baal/active/heuristics/heuristics.py [hunk 1]
```diff
@@ -68,6 +69,31 @@ def wrapper(self, probabilities):
...
+
+    Args:
+        data (tensor): A tensor of data.
+        dim (int): dimension to expand along.
+        index (tensor): tensor with the indices to gather.
+
+    Returns:
+        Tensor with the same shape as `index`.
+    """
+    max_shape = [max(dr, ir) for dr, ir in zip(data.shape, index.shape)]
```
Comment (BlackHC) [inline src/baal/active/heuristics/heuristics.py:84]: Could you add a reference to https://github.com/BlackHC/BatchBALD/blob/master/src/torch_utils.py ?

File: src/baal/active/heuristics/heuristics.py [hunk 2]
```diff
@@ -241,27 +272,70 @@ class BatchBALD(BALD):
...
+        probs_B_C = probs.reshape((-1, probs.shape[-1]))
+
+        # samples: Ni... x draw_per_xx
+        choices = torch.multinomial(probs_B_C,
+                                    num_samples=n_choices, replacement=True)
+
+        choices_b_M = choices.reshape(list(probs.shape[:-1]) + [n_choices])
+        return choices_b_M.long()
+
+    def _sample_from_history(self, probs, num_draw=1000):
```
Comment (BlackHC) [inline src/baal/active/heuristics/heuristics.py:299]: And a reference to https://github.com/BlackHC/BatchBALD/blob/master/src/joint_entropy/sampling.py ?

File: src/baal/active/heuristics/heuristics.py [hunk 3]
```diff
@@ -241,27 +272,70 @@ class BatchBALD(BALD):
...
-    def __init__(self, num_samples, shuffle_prop=0.0, reduction='none'):
+    def __init__(self, num_samples, num_draw=500, shuffle_prop=0.0, reduction='none'):
         self.epsilon = 1e-5
         self.num_samples = num_samples
+        self.num_draw = num_draw
         super().__init__(shuffle_prop=shuffle_prop,
                          reduction=reduction)
 
+    def _draw_choices(self, probs, n_choices):
+        """
```
Comment (BlackHC) [inline src/baal/active/heuristics/heuristics.py:283]: And a reference to https://github.com/BlackHC/BatchBALD/blob/master/src/torch_utils.py#L187 ?

File: src/baal/active/heuristics/heuristics.py [hunk 4]
```diff
@@ -275,29 +349,31 @@ def compute_score(self, predictions):
...
-        MAX_SELECTED = 16000
         MIN_SPREAD = 0.1
         COUNT = 0
         # Get conditional_entropies_B
         conditional_entropies_B = self._conditional_entropy(predictions)
         bald_out = super().compute_score(predictions)
-        history = [self.reduction(bald_out).argmax()]
+        # We start with the most uncertain sample according to BALD.
+        history = self.reduction(bald_out).argsort()[-1:].tolist()
         for step in range(2 * self.num_samples):
```
Comment (BlackHC) [inline src/baal/active/heuristics/heuristics.py:359]: Why 2*self.num_samples?

**Comment by Dref360** [comment]
Oh sorry about the references @BlackHC I added the one missing.

**Pull request merged**